File: fore-retrofit/src/main/java/co/early/fore/retrofit/ErrorHandler.java
package co.early.fore.retrofit;

import androidx.annotation.Nullable;
import okhttp3.Request;
import retrofit2.Response;

/**
 *
 * @param <F> Failure message class to be used throughout the app, such as an enum
 */
public interface ErrorHandler<F> {
    /**
     *
     * @param t throwable that caused the error, may be null
     * @param errorResponse error response from the server, the body of which may represent a custom error, may be null
     * @param customErrorClazz custom error class expected from the errorResponse, may be null
     * @param originalRequest in case it's needed (if you need access to the headers for example), may be null
     * @param <CE> class type of the custom error, may be null
     * @return the parsed error from the server
     */
    <CE extends MessageProvider<F>> F handleError(@Nullable Throwable t, @Nullable Response errorResponse, @Nullable Class<CE> customErrorClazz, @Nullable Request originalRequest);
}

File: lib/fore-net/src/commonMain/kotlin/co/early/fore/net/wrap/ErrorHandler.kt
package co.early.fore.net.wrap

import co.early.fore.net.MessageProvider
import kotlinx.serialization.KSerializer

/**
 *
 * @param <F> Failure message class to be used throughout the app (or the data layer in the case of
 * clean architecture) such as an enum or a data class
 */
interface ErrorHandler<F> {
    /**
     *
     * @param t throwable that caused the error
     * @param KSerializer serializer for the custom error class expected from the errorResponse, may be null
     * @param <CE> class type of the custom error if specified
     * @return the parsed error from the server
    </CE> */
    suspend fun <CE : MessageProvider<F>> handleError(t: Throwable, kSerializer: KSerializer<CE>?): F
}
